File: src-c/uart.c
// uart.c - UART Driver for Raspberry Pi 5 (PL011)

// Type definitions (replacing stdint.h)
typedef unsigned char uint8_t;
typedef unsigned int uint32_t;
typedef unsigned long uint64_t;

// Raspberry Pi 5 UART0 base address (BCM2712)
// Reference: pi5_os early_uart implementation
#define UART0_BASE 0x107d001000UL

// UART registers
#define UART0_DR     ((volatile uint32_t*)(UART0_BASE + 0x00))
#define UART0_FR     ((volatile uint32_t*)(UART0_BASE + 0x18))
#define UART0_IBRD   ((volatile uint32_t*)(UART0_BASE + 0x24))
#define UART0_FBRD   ((volatile uint32_t*)(UART0_BASE + 0x28))
#define UART0_LCRH   ((volatile uint32_t*)(UART0_BASE + 0x2C))
#define UART0_CR     ((volatile uint32_t*)(UART0_BASE + 0x30))
#define UART0_IMSC   ((volatile uint32_t*)(UART0_BASE + 0x38))
#define UART0_ICR    ((volatile uint32_t*)(UART0_BASE + 0x44))

// UART FR register bits
#define UART_FR_TXFF (1 << 5)  // Transmit FIFO full
#define UART_FR_RXFE (1 << 4)  // Receive FIFO empty

void uart_init(void) {
    // Pi5 UART is already initialized by bootloader
    // No initialization needed for early UART
    // Reference: pi5_os early_uart implementation
}

void uart_putc(char c) {
    // Wait until TX FIFO is not full
    while (*UART0_FR & UART_FR_TXFF);
    *UART0_DR = c;
}

void uart_puts(const char* str) {
    while (*str) {
        if (*str == '\n') {
            uart_putc('\r');
        }
        uart_putc(*str++);
    }
}

// Helper: print a 4-bit nibble as hex
static void uart_puthex_nibble(unsigned int v) {
    unsigned int nib = v & 0xF;
    if (nib < 10) uart_putc('0' + nib);
    else uart_putc('a' + (nib - 10));
}

// Print 32-bit value as 0xhhhhhhhh
void uart_puthex32(unsigned int v) {
    uart_puts("0x");
    for (int i = 7; i >= 0; --i) {
        unsigned int shift = i * 4;
        uart_puthex_nibble((v >> shift) & 0xF);
    }
}

// Print 64-bit value as 0xhhhhhhhhhhhhhhhh
void uart_puthex64(unsigned long v) {
    uart_puts("0x");
    for (int i = 15; i >= 0; --i) {
        unsigned long shift = (unsigned long)i * 4UL;
        unsigned int nib = (unsigned int)((v >> shift) & 0xFUL);
        if (nib < 10) uart_putc('0' + nib);
        else uart_putc('a' + (nib - 10));
    }
}

char uart_getc(void) {
    // Wait until RX FIFO is not empty
    while (*UART0_FR & UART_FR_RXFE);
    return (char)(*UART0_DR & 0xFF);
}

File: drivers/uart.rs
// uart.rs - UART Driver for Raspberry Pi 5

use core::ptr;

// BCM2712 UART base address - Pi5 specific
const UART0_BASE: usize = 0x107d001000;

// UART registers
const UART0_DR: *mut u32 = (UART0_BASE + 0x00) as *mut u32;
const UART0_FR: *mut u32 = (UART0_BASE + 0x18) as *mut u32;

// UART FR register bits
const UART_FR_TXFF: u32 = 1 << 5; // Transmit FIFO full
const UART_FR_RXFE: u32 = 1 << 4; // Receive FIFO empty

pub struct Uart;

impl Uart {
    pub fn new() -> Self {
        Uart
    }

    pub fn putc(&self, c: u8) {
        unsafe {
            // Wait until TX FIFO is not full
            while ptr::read_volatile(UART0_FR) & UART_FR_TXFF != 0 {}
            ptr::write_volatile(UART0_DR, c as u32);
        }
    }

    pub fn puts(&self, s: &str) {
        for byte in s.bytes() {
            if byte == b'\n' {
                self.putc(b'\r');
            }
            self.putc(byte);
        }
    }

    pub fn getc(&self) -> u8 {
        unsafe {
            // Wait until RX FIFO is not empty
            while ptr::read_volatile(UART0_FR) & UART_FR_RXFE != 0 {}
            (ptr::read_volatile(UART0_DR) & 0xFF) as u8
        }
    }

    pub fn try_getc(&self) -> Option<u8> {
        unsafe {
            if ptr::read_volatile(UART0_FR) & UART_FR_RXFE != 0 {
                None
            } else {
                Some((ptr::read_volatile(UART0_DR) & 0xFF) as u8)
            }
        }
    }
}

use core::fmt::{self, Write};

impl Write for Uart {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        self.puts(s);
        Ok(())
    }
}

// Wrapper for static UART access with interior mutability
pub struct UartWriter;

impl Write for UartWriter {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        UART.puts(s);
        Ok(())
    }
}

// Global UART instance
pub static UART: Uart = Uart;

#[macro_export]
macro_rules! print {
    ($($arg:tt)*) => {
        {
            use core::fmt::Write;
            let _ = write!($crate::drivers::uart::UartWriter, $($arg)*);
        }
    };
}

#[macro_export]
macro_rules! println {
    () => ($crate::print!("\n"));
    ($($arg:tt)*) => {
        {
            use core::fmt::Write;
            let _ = write!($crate::drivers::uart::UartWriter, $($arg)*);
            $crate::print!("\n");
        }
    };
}
